## Hugging Face NLP Course

### Behind the pipeline

![Diagram](https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/full_nlp_pipeline.svg)

- Use tokenizer to prepare the data for the processing with the model. Convert words, punctuation and special symbols to tokens.
- Converting sentences become sequences of tokens (much like the tokenizer in TensorFlow). The difference being that here the tokenizer is also responsible for padding and truncation.

In [1]:
from transformers import AutoTokenizer

/opt/conda/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

Once we have the tokenizer, we can directly pass our sentences to it and we’ll get back a dictionary that’s ready to feed to our model! The only thing left to do is to convert the list of input IDs to tensors.

In [3]:
raw_inputs = [
    "I've been waiting for a HuggingFace course my whole life.",
    "I hate this so much!",
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])}


The output itself is a dictionary containing two keys, `input_ids` and `attention_mask`. `input_ids` contains two rows of integers (one for each sentence) that are the unique identifiers of the tokens in each sentence. We’ll explain what the `attention_mask` is later in this chapter.

### Going through the model

We can download our pretrained model the same way we did with our tokenizer. 🤗 Transformers provides an AutoModel class which also has a from_pretrained() method:

In [4]:
from transformers import AutoModel

model = AutoModel.from_pretrained(checkpoint)

In [5]:
# checking the shape of the last hidden layer?
outputs = model(**inputs)
outputs.last_hidden_state.shape

torch.Size([2, 16, 768])

Note that the outputs of 🤗 Transformers models behave like namedtuples or dictionaries. You can access the elements by attributes (like we did) or by key (`outputs["last_hidden_state"]`), or even by index if you know exactly where the thing you are looking for is (`outputs[0]`).

#### The Model Diagram

![Diagram](https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/transformer_and_head.svg)

In this diagram, the model is represented by its embeddings layer and the subsequent layers. The embeddings layer converts each input ID in the tokenized input into a vector that represents the associated token. The subsequent layers manipulate those vectors using the attention mechanism to produce the final representation of the sentences.

There are many different architectures available in 🤗 Transformers, with each one designed around tackling a specific task. Here is a non-exhaustive list:

- *Model (retrieve the hidden states)
- *ForCausalLM
- *ForMaskedLM
- *ForMultipleChoice
- *ForQuestionAnswering
- *ForSequenceClassification
- *ForTokenClassification

and others 🤗

Different transformer architectures yield different classification heads. So to be able to classify sentences sentiment as *positive* or *negative*, we would use the `AutoModelForSequenceClassification` from the list above. Code below:

In [6]:
from transformers import AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
outputs = model(**inputs)

outputs, outputs.logits.shape

(SequenceClassifierOutput(loss=None, logits=tensor([[-1.5607,  1.6123],
         [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None),
 torch.Size([2, 2]))

Looking at the output above, we can now see a [2,2] shape. Each sentence we processed resulted in tensors for the two outputs (Positive/ Negative) like this:

In [7]:
outputs.logits

tensor([[-1.5607,  1.6123],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>)

Our model predicted [-1.5607, 1.6123] for the first sentence and [ 4.1692, -3.3464] for the second one. Those are not probabilities but logits, the raw, unnormalized scores outputted by the last layer of the model (seems to be using a variation of Adam optimizer). To be converted to probabilities, they need to go through a SoftMax layer for multiclass (in this case 2) classification.

All 🤗 Transformers models output the logits, as the loss function for training will generally fuse the last activation function, such as SoftMax, with the actual loss function, such as cross entropy.

In [8]:
import torch

predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
print(predictions), model.config.id2label

tensor([[4.0195e-02, 9.5980e-01],
        [9.9946e-01, 5.4418e-04]], grad_fn=<SoftmaxBackward0>)


(None, {0: 'NEGATIVE', 1: 'POSITIVE'})

### The Model

Creating a model without using the AutoModel class.

The AutoModel class and all of its relatives are actually simple wrappers over the wide variety of models available in the library. It’s a clever wrapper as it can automatically guess the appropriate model architecture for your checkpoint, and then instantiates a model with this architecture.

Creating a model from the default configuration initializes it with random values:

In [9]:
from transformers import BertConfig, BertModel

# Building the config
config = BertConfig()

# Building the model from the config, initialised with random values
model = BertModel(config)

config

BertConfig {
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.36.0",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}

The model can be used in this state, but it will output gibberish; it needs to be trained first. We could train the model from scratch on the task at hand, but as you saw in Chapter 1, this would require a long time and a lot of data, and it would have a non-negligible environmental impact. To avoid unnecessary and duplicated effort, it’s imperative to be able to share and reuse models that have already been trained.

Loading a Transformer model that is already trained is simple — we can do this using the from_pretrained() method:

In [10]:
# Loading pretrained model weights:

model = BertModel.from_pretrained("bert-base-cased")

config.json: 100%|██████████| 570/570 [00:00<00:00, 5.68MB/s]
model.safetensors: 100%|██████████| 436M/436M [00:06<00:00, 69.4MB/s] 


As you saw earlier, we could replace BertModel with the equivalent AutoModel class. We’ll do this from now on as this produces checkpoint-agnostic code; if your code works for one checkpoint, it should work seamlessly with another. This applies even if the architecture is different, as long as the checkpoint was trained for a similar task (for example, a sentiment analysis task).

In the code sample above we didn’t use BertConfig, and instead loaded a pretrained model via the bert-base-cased identifier. This is a model checkpoint that was trained by the authors of BERT themselves; you can find more details about it in its model [card](https://huggingface.co/bert-base-cased).

This model is now initialized with all the weights of the checkpoint. It can be used directly for inference on the tasks it was trained on, and it can also be fine-tuned on a new task. By training with pretrained weights rather than from scratch, we can quickly achieve good results.

> 🗒️ The weights have been downloaded and cached (so future calls to the from_pretrained() method won’t re-download them) in the cache folder, which defaults to ~/.cache/huggingface/transformers. You can customize your cache folder by setting the HF_HOME environment variable.

In [11]:
sequences = ["Hello!", "Cool.", "Nice!"]
tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

model_inputs = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")
outputs = model(**model_inputs)

outputs.last_hidden_state.shape

tokenizer_config.json: 100%|██████████| 29.0/29.0 [00:00<00:00, 200kB/s]
vocab.txt: 100%|██████████| 213k/213k [00:00<00:00, 902kB/s]
tokenizer.json: 100%|██████████| 436k/436k [00:00<00:00, 1.22MB/s]


torch.Size([3, 4, 768])

Here I still do not know what to do with the output of the model to convert it to a meaningful, human readable output. Let's revisit this soon.

For the time being we can save the model to finish this section:

In [52]:
model.save_pretrained("./data/models/my-bert-model")

In [53]:
new_model = AutoModel.from_pretrained("./data/models/my-bert-model")

### Tokenizers

Tokenizers are one of the core components of the NLP pipeline. They serve one purpose: to translate text into data that can be processed by the model. Models can only process numbers, so tokenizers need to convert our text inputs to numerical data. In this section, we’ll explore exactly what happens in the tokenization pipeline.

In NLP tasks, the data that is generally processed is raw text. Here’s an example of such text:

> Jim Henson was a puppeteer

Different types of tokenizers 
- simplest: word based.
- another: character based.
- most state of the art: subword based where root of the word is reused across different occurences of the word
    - For example: `tokenization` is split into `token` and `##ization` where `ization` is the suffix of the word denoted by `##`
    - This helps the model understand that token, tokenizing, tokens all have similar meaning
    - also words like modernization, organization, optimization are constructed in similar way as tokenization
 
Similar to AutoModel, the AutoTokenizer class will grab the proper tokenizer class in the library based on the checkpoint name, and can be used directly with any checkpoint:

In [12]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

tokenizer("Using a Transformer network is simple")

{'input_ids': [101, 7993, 170, 13809, 23763, 2443, 1110, 3014, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

You can save and load tokens very much the same way as saving and loading model checkpoints.

### Encoding

Tranlsating text to numbers to be prepared as model inputs is called **Encoding**. Encoding is done in two steps
- The tokenisation
- conversion to input IDs

As we’ve seen, the first step is to split the text into words (or parts of words, punctuation symbols, etc.), usually called **tokens**. There are multiple rules that can govern that process, which is why we need to instantiate the tokenizer using the name of the model, to make sure we use the same rules that were used when the model was pretrained.

The second step is to convert those tokens into numbers, so we can build a tensor out of them and feed them to the model. To do this, the tokenizer has a **vocabulary**, which is the part we download when we instantiate it with the `from_pretrained()`.

To demonstrate how the tokenizer works, we are going to explore its class methods separately. In most cases passing list of sentences directly to the `tokenizer()` would perform suffice as it would execute both of the below steps automatically.


In [13]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

sequence = "Using a Transformer network is simple"
tokens = tokenizer.tokenize(sequence)

# this only splits the input into tokens list but performs no conversion to ids.
print(tokens)

['Using', 'a', 'Trans', '##former', 'network', 'is', 'simple']


This tokenizer is a subword tokenizer: it splits the words until it obtains tokens that can be represented by its vocabulary. That’s the case here with transformer, which is split into two tokens: **trans** and **##former**.

In [9]:
# second step in the encoding is converting tokens to token_ids
ids = tokenizer.convert_tokens_to_ids(tokens)

print(ids)

[7993, 170, 13809, 23763, 2443, 1110, 3014]


**Decoding** is going the other way around: from vocabulary indices, we want to get a string. This can be done with the decode() method as follows:

In [60]:
decoded_string = tokenizer.decode(ids)
print(decoded_string)

Using a Transformer network is simple


### Models expect a batch of inputs

In the previous exercise you saw how sequences get translated into lists of numbers. Let’s convert this list of numbers to a tensor and send it to the model:


In [14]:
import torch

from transformers import AutoTokenizer, AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

sequence = "I've been waiting for a Huggingface course my whole life."

tokens = tokenizer.tokenize(sequence)
ids = tokenizer.convert_tokens_to_ids(tokens)

## IMPORTANT: below we must add another dimension [] as the model would expect a batch of sentences, not just one.
input_ids = torch.tensor([ids]) 

input_ids, model(input_ids)


(tensor([[ 1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,  2607,
           2026,  2878,  2166,  1012]]),
 SequenceClassifierOutput(loss=None, logits=tensor([[-2.7276,  2.8789]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None))

**Batching** is the act of sending multiple sentences through the model, all at once. If you only have one sentence, you can just build a batch by wrapping it in an array, thus adding a dimension.

Batching, however, introduces a new issue. What if different sentences have different word lenghts. We solve this via **padding**. Adding zeroes to the left or right of shorter sentences until we get the maximum sentence lenght. If the max sentence length is bigger than what the model can handle, we truncate.

**Attention Masks** are tensors with the exact same shape as the input IDs tensor, filled with 0s and 1s: 1s indicate the corresponding tokens should be attended to, and 0s indicate the corresponding tokens should not be attended to (i.e., they should be ignored by the attention layers of the model).

### Putting It All Together

🤗 Transformers API can handle all of the above processes for us with a high-level function that we’ll dive into here. When you call your tokenizer directly on the sentence, you get back inputs that are ready to pass through your model:

In [12]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

sequence = "I've been waiting for a HuggingFace course my whole life."
# high level function - does not need to be split into sub procedures like tokenize, 
# convert_tokens_to_ids, etc.
model_inputs = tokenizer(sequence)

# we can process number of sequences (sentences)

sequences = ["I've been waiting for a HuggingFace course my whole life.", "So have I!"]

# we can apply different padding rules

# Will pad the sequences up to the maximum sequence length
model_inputs = tokenizer(sequences, padding="longest")

# Will pad the sequences up to the model max length
# (512 for BERT or DistilBERT)
model_inputs = tokenizer(sequences, padding="max_length")

# Will pad the sequences up to the specified max length
model_inputs = tokenizer(sequences, padding="max_length", max_length=8)

# truncation

# Will truncate the sequences that are longer than the model max length
# (512 for BERT or DistilBERT)
model_inputs = tokenizer(sequences, truncation=True)

# Will truncate the sequences that are longer than the specified max length
model_inputs = tokenizer(sequences, max_length=8, truncation=True)

### Convert to different tensors or numpy

```python
sequences = ["I've been waiting for a HuggingFace course my whole life.", "So have I!"]

# Returns PyTorch tensors (Torch must be installed)
model_inputs = tokenizer(sequences, padding=True, return_tensors="pt")

# Returns TensorFlow tensors (TensorFlow must be installed)
model_inputs = tokenizer(sequences, padding=True, return_tensors="tf")

# Returns NumPy arrays (Numpy must be installed)
model_inputs = tokenizer(sequences, padding=True, return_tensors="np")
```

The **tokenizer** added the special word [CLS] *encoded as 101* at the beginning and the special word [SEP] *encoded as 102* at the end. This is because the model was pretrained with those, so to get the same results for inference we need to add them as well. Note that some models don’t add special words, or add different ones; models may also add these special words only at the beginning, or only at the end. In any case, the tokenizer knows which ones are expected and will deal with this for you.

In [15]:
sequence = "I've been waiting for a HuggingFace course my whole life."

model_inputs = tokenizer(sequence)
print(model_inputs["input_ids"])

[101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102]


### Wrapping up: From tokenizer to model

Now that we’ve seen all the individual steps the tokenizer object uses when applied on texts, let’s see one final time how it can handle multiple sequences (padding!), very long sequences (truncation!), and multiple types of tensors with its main API:

In [16]:
import torch
import numpy as np
from transformers import AutoTokenizer, AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
sequences = ["I've been waiting for a HuggingFace course my whole life.", 
             "So have I!", 
             "This is so cool!", 
             "I just hate the war.",
             "If I were any better than that, I would have guessed correctly."]

tokens = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")
with torch.no_grad():
    logits = model(**tokens).logits

predicted_index = np.array(logits).argmax(axis=1)
[model.config.id2label[key] for key in predicted_index]



['POSITIVE', 'POSITIVE', 'POSITIVE', 'NEGATIVE', 'NEGATIVE']

In [5]:
!pip install sentencepiece

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 37.7 MB/s eta 0:00:0000:01


In [17]:
import torch
import numpy as np
from transformers import AutoTokenizer, AutoModelForSequenceClassification

checkpoint = "cardiffnlp/twitter-xlm-roberta-base-sentiment"
tokenizer = AutoTokenizer.from_pretrained(checkpoint, use_fast=False)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
sequences = ["I've been waiting for a HuggingFace course my whole life.", 
             "So have I!", 
             "This is so cool!", 
             "I just hate the war.",
             "If I were any better than that, I would have guessed correctly."]

tokens = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")
with torch.no_grad():
    logits = model(**tokens).logits

predicted_index = np.array(logits).argmax(axis=1)
[model.config.id2label[key] for key in predicted_index]

config.json: 100%|██████████| 841/841 [00:00<00:00, 9.74MB/s]
sentencepiece.bpe.model: 100%|██████████| 5.07M/5.07M [00:00<00:00, 20.2MB/s]
special_tokens_map.json: 100%|██████████| 150/150 [00:00<00:00, 1.61MB/s]


ValueError: Couldn't instantiate the backend tokenizer from one of: 
(1) a `tokenizers` library serialization file, 
(2) a slow tokenizer instance to convert or 
(3) an equivalent slow tokenizer class to instantiate and convert. 
You need to have sentencepiece installed to convert a slow tokenizer to a fast one.

## Fine Tuning

Continuing with the example from the previous chapter, here is how we would train a sequence classifier on one batch in PyTorch:

In [18]:
import torch
from transformers import AdamW, AutoTokenizer, AutoModelForSequenceClassification

# Same as before
checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
sequences = [
    "I've been waiting for a HuggingFace course my whole life.",
    "This course is amazing!",
]
batch = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")

# This is new
batch["labels"] = torch.tensor([1, 1])

optimizer = torch.optim.AdamW(model.parameters())
loss = model(**batch).loss
loss.backward()
optimizer.step()

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


### Preprocessing Data with Torch

In this section we will use as an example the MRPC (Microsoft Research Paraphrase Corpus) dataset, introduced in a [paper](https://aclanthology.org/I05-5002.pdf) by William B. Dolan and Chris Brockett. The dataset consists of 5,801 pairs of sentences, with a label indicating if they are paraphrases or not (i.e., if both sentences mean the same thing). We’ve selected it for this chapter because it’s a small dataset, so it’s easy to experiment with training on it.

This is one of the 10 datasets composing the GLUE benchmark, which is an academic benchmark that is used to measure the performance of ML models across 10 different text classification tasks.

The 🤗 Datasets library provides a very simple command to download and cache a dataset on the Hub. We can download the MRPC dataset like this:

In [4]:
from datasets import load_dataset

raw_datasets = load_dataset("glue", "mrpc")
raw_datasets

DatasetDict({
    train: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 3668
    })
    validation: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 408
    })
    test: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 1725
    })
})

We can access each pair of sentences in our raw_datasets object by indexing, like with a dictionary:

In [5]:
raw_train_dataset = raw_datasets["train"]
raw_test_dataset = raw_datasets["test"]
raw_train_dataset[0], raw_train_dataset.features

({'sentence1': 'Amrozi accused his brother , whom he called " the witness " , of deliberately distorting his evidence .',
  'sentence2': 'Referring to him as only " the witness " , Amrozi accused his brother of deliberately distorting his evidence .',
  'label': 1,
  'idx': 0},
 {'sentence1': Value(dtype='string', id=None),
  'sentence2': Value(dtype='string', id=None),
  'label': ClassLabel(names=['not_equivalent', 'equivalent'], id=None),
  'idx': Value(dtype='int32', id=None)})

In [6]:
import pandas as pd

pd.DataFrame(raw_train_dataset[:10])

sentence1  \
0  Amrozi accused his brother , whom he called " ...   
1  Yucaipa owned Dominick 's before selling the c...   
2  They had published an advertisement on the Int...   
3  Around 0335 GMT , Tab shares were up 19 cents ...   
4  The stock rose $ 2.11 , or about 11 percent , ...   
5  Revenue in the first quarter of the year dropp...   
6  The Nasdaq had a weekly gain of 17.27 , or 1.2...   
7  The DVD-CCA then appealed to the state Supreme...   
8  That compared with $ 35.18 million , or 24 cen...   
9  Shares of Genentech , a much larger company wi...   

                                           sentence2  label  idx  
0  Referring to him as only " the witness " , Amr...      1    0  
1  Yucaipa bought Dominick 's in 1995 for $ 693 m...      0    1  
2  On June 10 , the ship 's owners had published ...      1    2  
3  Tab shares jumped 20 cents , or 4.6 % , to set...      0    3  
4  PG & E Corp. shares jumped $ 1.63 or 8 percent...      1    4  
5  With the scandal hanging over Stewart 's compa...      1    5  
6  The tech-laced Nasdaq Composite .IXIC rallied ...      0    6  
7  The DVD CCA appealed that decision to the U.S....      1    7  
8  Earnings were affected by a non-recurring $ 8 ...      0    8  
9  Shares of Xoma fell 16 percent in early trade ...      0   10

To preprocess the dataset, we need to convert the text to numbers the model can make sense of. As you saw in the previous chapter, this is done with a tokenizer. We can feed the tokenizer one sentence or a list of sentences, so we can directly tokenize all the first sentences and all the second sentences of each pair like this:

In [7]:
from transformers import AutoTokenizer

checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

We need to concat the two sentences and match them with the label 0 for non-similar and 1 for similar.

We need to handle the two sequences as a pair, and apply the appropriate preprocessing. Fortunately, the tokenizer can also take a pair of sequences and prepare it the way our BERT model expects:

here is with single record first:

In [24]:
inputs = tokenizer(raw_train_dataset[0]["sentence1"], raw_train_dataset[0]["sentence1"])

If we decode the IDs inside the input_ids back to words we get:

In [25]:
tokenizer.convert_ids_to_tokens(inputs["input_ids"])

['[CLS]',
 'am',
 '##ro',
 '##zi',
 'accused',
 'his',
 'brother',
 ',',
 'whom',
 'he',
 'called',
 '"',
 'the',
 'witness',
 '"',
 ',',
 'of',
 'deliberately',
 'di',
 '##stor',
 '##ting',
 'his',
 'evidence',
 '.',
 '[SEP]',
 'am',
 '##ro',
 '##zi',
 'accused',
 'his',
 'brother',
 ',',
 'whom',
 'he',
 'called',
 '"',
 'the',
 'witness',
 '"',
 ',',
 'of',
 'deliberately',
 'di',
 '##stor',
 '##ting',
 'his',
 'evidence',
 '.',
 '[SEP]']

To tokenize the complete data set we can use again the `tokenize` function. This however will load everything from the Dataset in RAM all at once during tokenization. It will also return a list of lists including columns that we do not need like `attention_mask`, `token_type_ids`


In [26]:
tokenized_dataset = tokenizer(
    raw_train_dataset["sentence1"],
    raw_train_dataset["sentence2"],
    padding=True,
    truncation=True,
)

In [28]:
# sample tokenized content.
tokenized_dataset["input_ids"][10:13] 

[[101,
  6094,
  2437,
  2009,
  6211,
  2005,
  10390,
  2000,
  22505,
  2037,
  13930,
  1999,
  10528,
  2457,
  2180,
  10827,
  2160,
  6226,
  1999,
  2233,
  1012,
  102,
  6094,
  2437,
  2009,
  6211,
  2005,
  10390,
  2000,
  22505,
  2037,
  13930,
  1999,
  10528,
  2457,
  2180,
  26203,
  1010,
  2160,
  6226,
  1999,
  2233,
  1998,
  2001,
  11763,
  2011,
  1996,
  2317,
  2160,
  1012,
  102,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0],
 [101,
  1996,
  17235,
  2850,
  4160,
  12490,
  5950,
  3445,
  2184,
  1012,
  6421,
  1010,
  2030,
  1014,
  1012,
  1021,
  3867,
  1010,
  2000,
  1015,
  1010,
  4868,
  2549,
  1012,
  6255,
  1012,
  102,
  1996,
  17235,
  2850,
  4160,
  12490,
  5950,
  1010,
  2440,
  1997,
  2974,
  15768,
  1010,
  2001,


**To keep the data as a dataset**, we will use the `Dataset.map()` method. This also allows us some extra flexibility, if we need more preprocessing done than just tokenization. The map() method works by applying a function on each element of the dataset, so let’s define a function that tokenizes our inputs:

In [10]:
# The tokenize function
def tokenize_function(example):
    return tokenizer(example["sentence1"], example["sentence2"], truncation=True)

Here is how we apply the tokenization function on all our datasets at once. We’re using batched=True in our call to map so the function is applied to multiple elements of our dataset at once, and not on each element separately. This allows for faster preprocessing.

In [11]:
tokenized_datasets = raw_datasets.map(tokenize_function, batched=True, batch_size=32)
tokenized_datasets["train"]["label"][:10]

[1, 0, 1, 0, 1, 1, 0, 1, 0, 0]

In [12]:
# Examine as a DataFrame

import pandas as pd

pd.DataFrame(tokenized_datasets["train"][:10])

sentence1  \
0  Amrozi accused his brother , whom he called " ...   
1  Yucaipa owned Dominick 's before selling the c...   
2  They had published an advertisement on the Int...   
3  Around 0335 GMT , Tab shares were up 19 cents ...   
4  The stock rose $ 2.11 , or about 11 percent , ...   
5  Revenue in the first quarter of the year dropp...   
6  The Nasdaq had a weekly gain of 17.27 , or 1.2...   
7  The DVD-CCA then appealed to the state Supreme...   
8  That compared with $ 35.18 million , or 24 cen...   
9  Shares of Genentech , a much larger company wi...   

                                           sentence2  label  idx  \
0  Referring to him as only " the witness " , Amr...      1    0   
1  Yucaipa bought Dominick 's in 1995 for $ 693 m...      0    1   
2  On June 10 , the ship 's owners had published ...      1    2   
3  Tab shares jumped 20 cents , or 4.6 % , to set...      0    3   
4  PG & E Corp. shares jumped $ 1.63 or 8 percent...      1    4   
5  With the scandal hanging over Stewart 's compa...      1    5   
6  The tech-laced Nasdaq Composite .IXIC rallied ...      0    6   
7  The DVD CCA appealed that decision to the U.S....      1    7   
8  Earnings were affected by a non-recurring $ 8 ...      0    8   
9  Shares of Xoma fell 16 percent in early trade ...      0   10   

                                           input_ids  \
0  [101, 2572, 3217, 5831, 5496, 2010, 2567, 1010...   
1  [101, 9805, 3540, 11514, 2050, 3079, 11282, 22...   
2  [101, 2027, 2018, 2405, 2019, 15147, 2006, 199...   
3  [101, 2105, 6021, 19481, 13938, 2102, 1010, 21...   
4  [101, 1996, 4518, 3123, 1002, 1016, 1012, 2340...   
5  [101, 6599, 1999, 1996, 2034, 4284, 1997, 1996...   
6  [101, 1996, 17235, 2850, 4160, 2018, 1037, 488...   
7  [101, 1996, 4966, 1011, 10507, 2050, 2059, 120...   
8  [101, 2008, 4102, 2007, 1002, 3486, 1012, 2324...   
9  [101, 6661, 1997, 4962, 10111, 2818, 1010, 103...   

                                      token_type_ids  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
5  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
6  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
7  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
8  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
9  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                      attention_mask  
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
1  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
2  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
3  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
4  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
5  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
6  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
7  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
8  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  
9  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...

You can even use multiprocessing when applying your preprocessing function with map() by passing along a `num_proc` argument. We didn’t do this here because the 🤗 Tokenizers library already uses multiple threads to tokenize our samples faster, but if you are not using a fast tokenizer backed by this library, this could speed up your preprocessing.

### Dynamic Padding

Dynamic padding is when we defer the padding from the tokenizer which would pad all sentences to the longest one in the dataset to a different class called DataCollator. This will pad sentences dynamically only to the longest in the **current batch**.

The function that is responsible for putting together samples inside a batch is called a **collate function**. It’s an argument you can pass when you build a DataLoader, the default being a function that will just convert your samples to PyTorch tensors and concatenate them (recursively if your elements are lists, tuples, or dictionaries). This won’t be possible in our case since the inputs we have won’t all be of the same size. We have deliberately postponed the padding, to only apply it as necessary on each batch and avoid having over-long inputs with a lot of padding. This will speed up training by quite a bit, but note that if you’re training on a TPU it can cause problems — TPUs prefer fixed shapes, even when that requires extra padding

To do this in practice, we have to define a collate function that will apply the correct amount of padding to the items of the dataset we want to batch together. Fortunately, the 🤗 Transformers library provides us with such a function via DataCollatorWithPadding. It takes a tokenizer when you instantiate it (to know which padding token to use, and whether the model expects padding to be on the left or on the right of the inputs) and will do everything you need:

In [33]:
from transformers import DataCollatorWithPadding

data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

samples = tokenized_datasets["train"][:10]
samples = {k: v for k, v in samples.items() if k not in ["idx", "sentence1", "sentence2"]}
[len(x) for x in samples["input_ids"]]

[50, 59, 47, 67, 59, 50, 62, 32, 45, 60]

In [34]:
batch = data_collator(samples)
{k: v.shape for k, v in batch.items()}

You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


{'input_ids': torch.Size([10, 67]),
 'token_type_ids': torch.Size([10, 67]),
 'attention_mask': torch.Size([10, 67]),
 'labels': torch.Size([10])}

### Fine-tuning a model with the Trainer API

🤗 Transformers provides a Trainer class to help you fine-tune any of the pretrained models it provides on your dataset.

Once you’ve done all the data preprocessing work in the last section, you have just a few steps left to define the Trainer. The hardest part is likely to be preparing the environment to run Trainer.train(), as it will run very slowly on a CPU. If you don’t have a GPU set up, you can get access to free GPUs or TPUs on Google Colab.

The code examples below assume you have already executed the examples in the previous section. Here is a short summary recapping what you need:

In [70]:
from datasets import load_dataset
from transformers import AutoTokenizer, DataCollatorWithPadding

raw_datasets = load_dataset("glue", "mrpc")

checkpoint = "bert-base-uncased"

tokenizer = AutoTokenizer.from_pretrained(checkpoint)

tokenized_datasets = raw_datasets.map(tokenize_function, batched=True, batch_size=32)

data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

### The Training

The first step before we can define our Trainer is to define a TrainingArguments class that will contain all the hyperparameters the Trainer will use for training and evaluation. The only argument you have to provide is a directory where the trained model will be saved, as well as the checkpoints along the way. For all the rest, you can leave the defaults, which should work pretty well for a basic fine-tuning.

In [36]:
from transformers import TrainingArguments
from transformers import AutoModelForSequenceClassification
from transformers import Trainer

training_args = TrainingArguments("./data/test-trainer", num_train_epochs=4)

model = AutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)
trainer = Trainer(
    model,
    training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    data_collator=data_collator,
    tokenizer=tokenizer,
)

trainer.train()

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Checkpoint destination directory ./data/test-trainer/checkpoint-500 already exists and is non-empty.Saving will proceed but saved results may be invalid.
Checkpoint destination directory ./data/test-trainer/checkpoint-1000 already exists and is non-empty.Saving will proceed but saved results may be invalid.
Checkpoint destination directory ./data/test-trainer/checkpoint-1500 already exists and is non-empty.Saving will proceed but saved results may be invalid.


TrainOutput(global_step=1836, training_loss=0.2538219163100964, metrics={'train_runtime': 49.252, 'train_samples_per_second': 297.897, 'train_steps_per_second': 37.278, 'total_flos': 541242051981840.0, 'train_loss': 0.2538219163100964, 'epoch': 4.0})

This will start the fine-tuning and report the training loss every 500 steps. It won’t, however, tell you how well (or badly) your model is performing. This is because:

1. We didn’t tell the Trainer to evaluate during training by setting evaluation_strategy to either "steps" (evaluate every eval_steps) or "epoch" (evaluate at the end of each epoch).
2. We didn’t provide the Trainer with a compute_metrics() function to calculate a metric during said evaluation (otherwise the evaluation would just have printed the loss, which is not a very intuitive number).

Let’s see how we can build a useful compute_metrics() function and use it the next time we train. The function must take an EvalPrediction object (which is a named tuple with a predictions field and a label_ids field) and will return a dictionary mapping strings to floats (the strings being the names of the metrics returned, and the floats their values). 

To get some predictions from our model, we can use the Trainer.predict() command:

In [56]:
predictions = trainer.predict(tokenized_datasets["test"])
print(predictions.predictions.shape, predictions.label_ids.shape)

(1725, 2) (1725,)


As you can see, predictions is a two-dimensional array with shape n x 2 (n being the number of elements in the dataset we used). 2 are the logits for each element of the dataset we passed to predict(). To get the index of the prediction, we will use Numpy's argmax function. See `return_pred_index` function below.

In [67]:
# in logits and converted with argmax
predictions.predictions[97], np.array(predictions.predictions[97]).argmax(axis=-1)

(array([-3.7710178,  4.354563 ], dtype=float32), 1)

In [40]:
import numpy as np
def return_pred_index(logits):
    return np.array(logits).argmax(axis=-1)

The output of the `predict()` method is another named tuple with three fields: `predictions`, `label_ids`, and `metrics`. 

In [59]:
tokenized_datasets["test"]["label"][100:110], return_pred_index(predictions.predictions[100:110])

([1, 1, 1, 1, 1, 0, 1, 1, 1, 1], array([0, 1, 1, 1, 1, 0, 0, 1, 1, 1]))

In [62]:
list(zip([tokenized_datasets["test"]["sentence1"][116]], [tokenized_datasets["test"]["sentence2"][116]]))

[("NASA plans to follow-up the rovers ' missions with additional orbiters and landers before launching a long-awaited sample-return flight .",
  'NASA plans to explore the Red Planet with ever more sophisticated robotic orbiters and landers .')]

### Evaluation with Metrics

In [64]:
!pip install evaluate
import evaluate

preds = return_pred_index(predictions.predictions)

metric = evaluate.load("glue", "mrpc")
metric.compute(predictions=preds, references=predictions.label_ids)

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 6.1 MB/s eta 0:00:00


{'accuracy': 0.8527536231884058, 'f1': 0.8934563758389261}

Wrapping everything together, we get our `compute_metrics()` function:

In [71]:
def compute_metrics(eval_preds):
    metric = evaluate.load("glue", "mrpc")
    logits, labels = eval_preds
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)

And to see it used in action to report metrics at the end of each epoch, here is how we define a new Trainer with this `compute_metrics()` function:

In [72]:
training_args = TrainingArguments("./data/test-trainer", evaluation_strategy="epoch", num_train_epochs=4)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)

trainer = Trainer(
    model,
    training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["validation"],
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics,
)
trainer.train()

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Checkpoint destination directory test-trainer/checkpoint-500 already exists and is non-empty.Saving will proceed but saved results may be invalid.
Checkpoint destination directory test-trainer/checkpoint-1000 already exists and is non-empty.Saving will proceed but saved results may be invalid.
Checkpoint destination directory test-trainer/checkpoint-1500 already exists and is non-empty.Saving will proceed but saved results may be invalid.


TrainOutput(global_step=1836, training_loss=0.5221048076687815, metrics={'train_runtime': 53.3629, 'train_samples_per_second': 274.947, 'train_steps_per_second': 34.406, 'total_flos': 541242051981840.0, 'train_loss': 0.5221048076687815, 'epoch': 4.0})

The Trainer will work out of the box on multiple GPUs or TPUs and provides lots of options, like mixed-precision training (use fp16 = True in your training arguments). We will go over everything it supports in Chapter 10.

This concludes the introduction to fine-tuning using the Trainer API. An example of doing this for most common NLP tasks will be given in Chapter 7, but for now let’s look at how to do the same thing in pure PyTorch.

### Full Training with PyTorch

Now we’ll see how to achieve the same results as we did in the last section without using the Trainer class. Again, we assume you have done the data processing in section 2. Here is a short summary covering everything you will need

In [1]:
from datasets import load_dataset
from transformers import AutoTokenizer, DataCollatorWithPadding

raw_datasets = load_dataset("glue", "mrpc")
checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)


def tokenize_function(example):
    return tokenizer(example["sentence1"], example["sentence2"], truncation=True)


tokenized_datasets = raw_datasets.map(tokenize_function, batched=True)
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)


/opt/conda/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
from transformers import AutoModelForSequenceClassification

model = AutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)
model.config

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


BertConfig {
  "_name_or_path": "bert-base-uncased",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.36.0",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}

#### Prepare for training

Before actually writing our training loop, we will need to define a few objects. The first ones are the dataloaders we will use to iterate over batches. But before we can define those dataloaders, we need to apply a bit of postprocessing to our tokenized_datasets, to take care of some things that the Trainer did for us automatically. Specifically, we need to:
- Remove the columns corresponding to values the model does not expect (like the sentence1 and sentence2 columns).
- Rename the column label to labels (because the model expects the argument to be named labels).
- Set the format of the datasets so they return PyTorch tensors instead of lists.



In [3]:
tokenized_datasets = tokenized_datasets.remove_columns(["sentence1", "sentence2", "idx"])
tokenized_datasets = tokenized_datasets.rename_column("label", "labels")
tokenized_datasets.set_format("torch")
tokenized_datasets["train"].column_names

['labels', 'input_ids', 'token_type_ids', 'attention_mask']

Now that this is done, we can easily define our dataloaders:

In [4]:
from torch.utils.data import DataLoader

train_dataloader = DataLoader(
    tokenized_datasets["train"], shuffle=True, batch_size=32, collate_fn=data_collator
)
eval_dataloader = DataLoader(
    tokenized_datasets["validation"], batch_size=32, collate_fn=data_collator
)

Check the dataloaders that there is not mitake

In [5]:
for batch in train_dataloader:
    break
{k: v.shape for k, v in batch.items()}

You're using a BertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


{'labels': torch.Size([32]),
 'input_ids': torch.Size([32, 83]),
 'token_type_ids': torch.Size([32, 83]),
 'attention_mask': torch.Size([32, 83])}

Loading the model from pretrained the same way as we did for the trainer class

In [6]:
from transformers import AutoModelForSequenceClassification
import torch

model = AutoModelForSequenceClassification.from_pretrained(checkpoint, num_labels=2)

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [7]:
outputs = model(**batch)
outputs.logits.shape #batch of 32, two labels (POSITIVE, NEGATIVE)

torch.Size([32, 2])

In [8]:
outputs.loss # loss is calculated since the dataloader presents a label to compare

tensor(0.6656, grad_fn=<NllLossBackward0>)

We’re almost ready to write our training loop! We’re just missing two things: an optimizer and a learning rate scheduler. Since we are trying to replicate what the Trainer was doing by hand, we will use the same defaults. The optimizer used by the Trainer is AdamW, which is the same as Adam, but with a twist for weight decay regularization (see [“Decoupled Weight Decay Regularization”](https://arxiv.org/abs/1711.05101) by Ilya Loshchilov and Frank Hutter)

In [9]:
optimizer = torch.optim.Adam(model.parameters(), lr=5e-5)

Finally, the learning rate scheduler used by default is just a linear decay from the maximum value (5e-5) to 0. To properly define it, we need to know the number of training steps we will take, which is the number of epochs we want to run multiplied by the number of training batches (which is the length of our training dataloader). The Trainer uses three epochs by default, so we will follow that:

In [10]:
from transformers import get_scheduler

num_epochs = 3
num_training_steps = num_epochs * len(train_dataloader)
lr_scheduler = get_scheduler(
    "linear",
    optimizer=optimizer,
    num_warmup_steps=0,
    num_training_steps=num_training_steps,
)
print(num_training_steps)

345


One last thing: we will want to use the GPU if we have access to one (on a CPU, training might take several hours instead of a couple of minutes). To do this, we define a device we will put our model and our batches on:

In [20]:
device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
model.to(device)

BertForSequenceClassification(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12,

In [13]:
from tqdm.auto import tqdm

progress_bar = tqdm(range(num_training_steps))

model.train()
for epoch in range(num_epochs):
    for batch in train_dataloader:
        batch = {k: v.to(device) for k, v in batch.items()}
        outputs = model(**batch)
        loss = outputs.loss
        loss.backward()

        optimizer.step()
        lr_scheduler.step()
        optimizer.zero_grad()
        progress_bar.update(1)

100%|██████████| 345/345 [00:30<00:00, 19.78it/s]

#### Some Testing - Its a little wierd

In [71]:

model.eval()
test = dict()
for index, batch in enumerate(eval_dataloader):
    if index == 8:
        test = {k: v.to(device) for k, v in batch.items()}
        break
    
test

{'labels': tensor([1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0,
         1, 1, 1, 0, 0, 1, 1, 1], device='cuda:0'),
 'input_ids': tensor([[  101, 24547,  1011,  ...,  9317,  1012,   102],
         [  101,  6005,  1010,  ...,  1012,   102,     0],
         [  101,  1999,  1996,  ...,     0,     0,     0],
         ...,
         [  101,  1000,  2057,  ...,     0,     0,     0],
         [  101,  2720,  1012,  ...,     0,     0,     0],
         [  101,  1996,  2260,  ...,     0,     0,     0]], device='cuda:0'),
 'token_type_ids': tensor([[0, 0, 0,  ..., 1, 1, 1],
         [0, 0, 0,  ..., 1, 1, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         ...,
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0],
         [0, 0, 0,  ..., 0, 0, 0]], device='cuda:0'),
 'attention_mask': tensor([[1, 1, 1,  ..., 1, 1, 1],
         [1, 1, 1,  ..., 1, 1, 0],
         [1, 1, 1,  ..., 0, 0, 0],
         ...,
         [1, 1, 1,  ..., 0, 0, 0],
         [1, 1, 1,  ...,

In [72]:
from torch import Tensor
with torch.no_grad():
    output = model(**test)
list(zip(return_pred_index(Tensor.cpu(output.logits)), np.array(Tensor.cpu(test["labels"]))))

[(1, 1),
 (1, 1),
 (1, 1),
 (0, 0),
 (0, 1),
 (1, 0),
 (1, 1),
 (1, 1),
 (1, 0),
 (1, 1),
 (0, 1),
 (0, 0),
 (0, 0),
 (1, 1),
 (0, 0),
 (1, 1),
 (1, 1),
 (1, 1),
 (1, 1),
 (1, 1),
 (1, 1),
 (1, 1),
 (1, 1),
 (0, 0),
 (1, 1),
 (1, 1),
 (1, 1),
 (0, 0),
 (1, 0),
 (1, 1),
 (1, 1),
 (1, 1)]